In [4]:
#pip install nbconvert pip install pyppeteer

In [2]:
#--allow-chromium-download

# Practicing NumPy

## 2.0 
Import numpy under the alias `np`.

In [2]:
# The code goes here
import numpy as np

## 2.1 

Create the following arrays:

* Create an array of 10 ones.

* Create an array of the integers 1 to 20.

* Create a 5 x 5 matrix of ones with a dtype int.

In [3]:
#Code comes here

ones_array = np.tile(1, 10)
print(ones_array)

integers_array = np.arange(1, 21)
print(ones_array)

matrix_array = np.ones((1,5,5), dtype="int")
print(matrix_array)
print(matrix_array.dtype)



[1 1 1 1 1 1 1 1 1 1]
[1 1 1 1 1 1 1 1 1 1]
[[[1 1 1 1 1]
  [1 1 1 1 1]
  [1 1 1 1 1]
  [1 1 1 1 1]
  [1 1 1 1 1]]]
int64


## 2.2 
Use numpy to:
1. Create an 3D matrix of 3 x 3 x 3 full of random numbers drawn from a standard normal distribution (hint: `np.random.randn()`)
2. Reshape the above array into shape (27,)

In [9]:
# Code comes here

matrix_3d = np.random.randn(3, 3, 3)
print(matrix_3d)

array_1d = matrix_3d.reshape(27)
print(array_1d)

[[[-0.83402991  0.91627687  0.3353395 ]
  [ 0.6744778   0.82175196 -0.26346026]
  [ 0.07874084  0.15876063 -1.55920113]]

 [[ 1.40179103 -1.31032094  0.57482221]
  [-0.06139339  1.65288932  0.94966123]
  [-1.07332913 -0.30954147 -0.51433162]]

 [[-0.13282349 -0.84731627  1.53142949]
  [ 1.07619243 -1.0031597   0.92008042]
  [ 1.53441706  0.24554827  0.81779749]]]
[-0.83402991  0.91627687  0.3353395   0.6744778   0.82175196 -0.26346026
  0.07874084  0.15876063 -1.55920113  1.40179103 -1.31032094  0.57482221
 -0.06139339  1.65288932  0.94966123 -1.07332913 -0.30954147 -0.51433162
 -0.13282349 -0.84731627  1.53142949  1.07619243 -1.0031597   0.92008042
  1.53441706  0.24554827  0.81779749]


# 2.3 

Create an array of 20 linearly spaced numbers between 1 and 10.

In [11]:
# Code comes here

arr20 = np.linspace(1, 10, 20)
print(arr20)

[ 1.          1.47368421  1.94736842  2.42105263  2.89473684  3.36842105
  3.84210526  4.31578947  4.78947368  5.26315789  5.73684211  6.21052632
  6.68421053  7.15789474  7.63157895  8.10526316  8.57894737  9.05263158
  9.52631579 10.        ]


## 2.4

Run the following code to create an array of shape 4 x 4 and then use indexing to produce the outputs shown below.

In [22]:
import numpy as np

In [6]:
a = np.arange(1, 26).reshape(5, -1)
print(a)

[[ 1  2  3  4  5]
 [ 6  7  8  9 10]
 [11 12 13 14 15]
 [16 17 18 19 20]
 [21 22 23 24 25]]


```
array([[ 9, 10],
       [14, 15],
       [19, 20],
       [24, 25]])
```

In [25]:
reshapearray = a[1:5, 3:5]
print(reshapearray)

[[ 9 10]
 [14 15]
 [19 20]
 [24 25]]


```python
array([ 6,  7,  8,  9, 10])
```

In [26]:
reshapearrary_2 = a[1,0:5]
print(reshapearrary_2)

[ 6  7  8  9 10]


## 2.5 

Calculate the sum of all the numbers in `a`.

In [27]:
# Code comes here

total_sum = a.sum()
print(total_sum)

325


## 2.6 

Calculate the sum of each row in `a`.

In [29]:
# Code comes here

row_sum = a.sum(axis=1)
print(row_sum)

[ 15  40  65  90 115]


## 2.7 

Extract all values of `a` greater than the mean of `a` (hint: use a boolean mask).

In [30]:
# Code comes here
a = np.arange(1, 26).reshape(5, 5) 
print(a)

[[ 1  2  3  4  5]
 [ 6  7  8  9 10]
 [11 12 13 14 15]
 [16 17 18 19 20]
 [21 22 23 24 25]]


In [7]:
mean_a = a.mean() #before we use the boolean mask we need to know the mean
print(mean_a)

13.0


In [37]:
booleanmask = a > mean_a #now we can create the mean. 
print(booleanmask)

[[False False False False False]
 [False False False False False]
 [False False False  True  True]
 [ True  True  True  True  True]
 [ True  True  True  True  True]]


In [39]:
above_mean = a[booleanmask] #now we filter to find all the values greater than the mean. 
print("\Values greater than the mean:")
print(above_mean)

\Values greater than the mean:
[14 15 16 17 18 19 20 21 22 23 24 25]


# Practicing Pandas

## 2.8

In this we'll be investigating the carbon footprint of different foods. 

We'll be leveraging a dataset compiled by [Kasia Kulma](https://r-tastic.co.uk/post/from-messy-to-tidy/) and contributed to [R's Tidy Tuesday project](https://github.com/rfordatascience/tidytuesday).

Start by importing pandas with the alias `pd`.

In [24]:
# Codes comes here
import pandas as pd

## 2.9 

The dataset we'll be working with has the following columns:

|column      |description |
|:-------------|:-----------|
|country       | Country Name |
|food_category | Food Category |
|consumption   | Consumption (kg/person/year) |
|co2_emmission | Co2 Emission (Kg CO2/person/year) |


Import the dataset as a dataframe named `df` from this url: <https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-02-18/food_consumption.csv>

In [25]:
# Code comes here

url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-02-18/food_consumption.csv"

carbonfootprint_df = pd.read_csv(url)

carbonfootprint_df.head(10)

country             food_category  consumption  co2_emmission
0  Argentina                      Pork        10.51          37.20
1  Argentina                   Poultry        38.66          41.53
2  Argentina                      Beef        55.48        1712.00
3  Argentina               Lamb & Goat         1.56          54.63
4  Argentina                      Fish         4.36           6.96
5  Argentina                      Eggs        11.39          10.46
6  Argentina        Milk - inc. cheese       195.08         277.87
7  Argentina  Wheat and Wheat Products       103.11          19.66
8  Argentina                      Rice         8.77          11.22
9  Argentina                  Soybeans         0.00           0.00

## 2.10 

How many rows and columns are there in the dataframe?

In [26]:
#Code comes here
carbonfootprint_df.shape #the .shape function tells you the number of rows and coloums.

(1430, 4)

In this case, there are 1430 rows and 4 columns

## 2.11

What is the mean `co2_emission` of the whole dataset?

In [27]:
# Code comes here
#the .mean value tells us the mean
mean_co2 = carbonfootprint_df["co2_emmission"].mean()
print("The mean CO2 emission of the dataset is:")
print(mean_co2)

The mean CO2 emission of the dataset is:
74.383993006993


## 2.12

What is the maximum `co2_emmission` in the dataset and which food type and country does it belong to?

In [28]:
# Code comes here
#the .max function tells us the highest value
max_co2 = carbonfootprint_df["co2_emmission"].max()
print("The max CO2 emmissions of the dataset is:")
print(max_co2)

The max CO2 emmissions of the dataset is:
1712.0


## 2.13

How many countries produce more than 1000 Kg CO2/person/year for at least one food type?

In [29]:
#code comes here
#First we need to filter for countries with co2 emissions higher than 1000 in the dataset:
high_co2 = carbonfootprint_df[carbonfootprint_df["co2_emmission"] > 1000] 
print(high_co2)

       country food_category  consumption  co2_emmission
2    Argentina          Beef        55.48        1712.00
13   Australia          Beef        33.86        1044.85
57         USA          Beef        36.24        1118.29
90      Brazil          Beef        39.25        1211.17
123    Bermuda          Beef        33.15        1022.94


In [30]:
#Now we can the number of individual countries by using the .nunique function
num_countries = high_co2["country"].nunique()
print("The number of countries that produce over 1000 kg CO2/person for at least one food type is:")
print(num_countries)

The number of countries that produce over 1000 kg CO2/person for at least one food type is:
5


## 2.14

Which country consumes the least amount of beef per person per year?


In [31]:
#first I need to select only the rows for beef:
beef_df = carbonfootprint_df[carbonfootprint_df["food_category"] == "Beef"]

#now I need to find the lowest consumption:
lowest_beef = beef_df["consumption"].min()

#now I can find the country with the lowest beef consumption. 
country_min_beef = beef_df[beef_df["consumption"] == lowest_beef]["country"]

#now I can print the result
print("The country that consumes the lowest amount of beef is:")
print(country_min_beef.tolist())

The country that consumes the lowest amount of beef is:
['Liberia']


## 2.15

What is the total emissions of all the meat products (Pork, Poultry, Fish, Lamb & Goat, Beef) in the dataset combined?

In [32]:
# code comes here

#first I need to create a line to define what meat products are:
meat_products = ["Beef", "Pork", "Fish", "Poultry", "Lamb & Goat"]

#now I can filter the dataset for meat products. 
# I do this by creating a boolean mask for rows in the coloum food_category, filtering for meat_products.
meat_df = carbonfootprint_df[carbonfootprint_df["food_category"].isin(meat_products)]

# Now I can sum the CO2 emissions of the meat_df
total_meat_emissions = meat_df["co2_emmission"].sum()

#finally, I need to print the result.
print("Total meat emissions:")
print(total_meat_emissions)

Total meat emissions:
74441.13


# 2.16 

What is the total emissions of all other (non-meat) products in the dataset combined?

To do this exercise, we can subtract the number of total meat emissions, found in 2.15 from the total emissions for the dataset.


In [58]:
#this line sums all the CO2 emmissions for all the food categories
Foodfootprint = carbonfootprint_df.groupby("food_category")["co2_emmission"].sum()
print(Foodfootprint)

food_category
Beef                        48633.26
Eggs                          974.95
Fish                         3588.22
Lamb & Goat                 11837.38
Milk - inc. cheese          23290.00
Nuts inc. Peanut Butter       951.99
Pork                         7419.11
Poultry                      2963.16
Rice                         4886.91
Soybeans                       50.35
Wheat and Wheat Products     1773.78
Name: co2_emmission, dtype: float64


In [57]:
Foodfootprint.sum() #this lines sums all the rows from above in order to get a total. 

106369.11000000002

In [65]:
#Now we can subtract the meat emissions from the total emissions to get the non-meat emissions
nonmeat_emissions = Foodfootprint.sum()-total_meat_emissions

print("The total non-meat CO2 emissions for the dataset:")
print(nonmeat_emissions)

The total non-meat CO2 emissions for the dataset:
31927.98000000001


# Final Challenge



In [66]:
import pandas as pd

In [70]:
#first I am uploading the file
df_worldcities = pd.read_csv('/arcgis/world_cities.csv')

df_worldcities.head() #Here I am checking the first 5 rows to explore my data and see that that the upload worked properly

city    country    pop    lat    lon  capital
0  'Abasan al-Jadidah  Palestine   5629  31.31  34.34        0
1  'Abasan al-Kabirah  Palestine  18999  31.32  34.35        0
2        'Abdul Hakim   Pakistan  47788  30.55  72.11        0
3  'Abdullah-as-Salam     Kuwait  21817  29.36  47.98        0
4               'Abud  Palestine   2456  32.03  35.07        0

In [80]:
#Now I am creating a new coloum by transofrming the pop column to population in millions.
df_worldcities["pop_m"] = df_worldcities["pop"]/1000000
df_worldcities.head()

city    country    pop    lat    lon  capital     pop_m
0  'Abasan al-Jadidah  Palestine   5629  31.31  34.34        0  0.005629
1  'Abasan al-Kabirah  Palestine  18999  31.32  34.35        0  0.018999
2        'Abdul Hakim   Pakistan  47788  30.55  72.11        0  0.047788
3  'Abdullah-as-Salam     Kuwait  21817  29.36  47.98        0  0.021817
4               'Abud  Palestine   2456  32.03  35.07        0  0.002456

In [81]:
#Now I am removing the original pop column
df_worldcities = df_worldcities.drop("pop", axis=1) #axis 1 indicates columns not rows. 
df_worldcities.head()

city    country    lat    lon  capital     pop_m
0  'Abasan al-Jadidah  Palestine  31.31  34.34        0  0.005629
1  'Abasan al-Kabirah  Palestine  31.32  34.35        0  0.018999
2        'Abdul Hakim   Pakistan  30.55  72.11        0  0.047788
3  'Abdullah-as-Salam     Kuwait  29.36  47.98        0  0.021817
4               'Abud  Palestine  32.03  35.07        0  0.002456

In [86]:
df_worldcities.head(150)

city      country    lat    lon  capital     pop_m
0    'Abasan al-Jadidah    Palestine  31.31  34.34        0  0.005629
1    'Abasan al-Kabirah    Palestine  31.32  34.35        0  0.018999
2          'Abdul Hakim     Pakistan  30.55  72.11        0  0.047788
3    'Abdullah-as-Salam       Kuwait  29.36  47.98        0  0.021817
4                 'Abud    Palestine  32.03  35.07        0  0.002456
..                  ...          ...    ...    ...      ...       ...
145           Aalestrup      Denmark  56.70   9.50        0  0.002784
146            Aalsmeer  Netherlands  52.27   4.76        0  0.023067
147               Aalst      Belgium  50.93   4.03        0  0.077876
148              Aalten  Netherlands  51.94   6.57        0  0.019283
149              Aalter      Belgium  51.08   3.45        0  0.018857

[150 rows x 6 columns]

Now I will choose a city that starts with the same letter as my name. I have chosen my city to be Cape Town, because my name is Chiara.

In [92]:
#I will create a boolean mask to only choose the rows where the the city is Cape Town. 
capetown_df = df_worldcities[df_worldcities['city'] == 'Cape Town'] 

print(capetown_df) #checking to see if it worked

           city       country    lat    lon  capital     pop_m
6418  Cape Town  South Africa -33.93  18.46        0  3.546429


Now I will select the five biggest cities from South Africa. 

In [95]:
#First I will filter for South Africa
SA_df = df_worldcities[df_worldcities['country'] == 'South Africa']

print(SA_df)

               city       country    lat    lon  capital     pop_m
387        Adelaide  South Africa -32.70  26.29        0  0.012222
869        Alberton  South Africa -26.33  28.13        0  0.209001
1015          Alice  South Africa -32.79  26.83        0  0.009796
1035   Aliwal North  South Africa -30.69  26.71        0  0.046991
1047     Allanridge  South Africa -27.74  26.67        0  0.032798
...             ...           ...    ...    ...      ...       ...
41796     Witrivier  South Africa -25.31  31.02        0  0.016286
41843  Wolmaranstad  South Africa -27.20  25.98        0  0.051426
41876     Worcester  South Africa -33.64  19.44        0  0.133113
42727       Zastron  South Africa -30.29  27.08        0  0.012474
42773       Zeerust  South Africa -25.54  26.08        0  0.017958

[226 rows x 6 columns]


In [100]:
#Now I will sort the values by the population column

SA_df = SA_df.sort_values(by= 'pop_m', ascending=False)

#This shows me the first 5 values
SA_df.head()

city       country    lat    lon  capital     pop_m
6418      Cape Town  South Africa -33.93  18.46        0  3.546429
9999         Durban  South Africa -29.87  30.99        0  3.244028
16390  Johannesburg  South Africa -26.19  28.04        0  2.091491
35953        Soweto  South Africa -26.28  27.84        0  1.766824
29838      Pretoria  South Africa -25.73  28.22        1  1.687779

In [102]:
# Now I will create a new dataframe for the top 5 cities in South Africa

Top5SA_df = SA_df.head(5)

print("The top 5 cities by population in South Africa")
print(Top5SA_df)

The top 5 cities by population in South Africa
               city       country    lat    lon  capital     pop_m
6418      Cape Town  South Africa -33.93  18.46        0  3.546429
9999         Durban  South Africa -29.87  30.99        0  3.244028
16390  Johannesburg  South Africa -26.19  28.04        0  2.091491
35953        Soweto  South Africa -26.28  27.84        0  1.766824
29838      Pretoria  South Africa -25.73  28.22        1  1.687779
